### PR TITLE
Override tokenize.ensure-deterministic config flag

### PR DESCRIPTION
**Pull request by crusaderky**
- Add optional parameter to `tokenize(ensure_deterministic=True)` to force raising exception
- Can now run `test_tokenize.py` with either value for the config variable
- Subclass RuntimeError to restrict exception catching (we're calling pickle internally, which is very complex)
- Better tests for the `_seen` variable

### CODE REVIEW

File: dask/base.py [hunk 1]
```diff
@@ -1007,29 +1007,103 @@ def persist(*args, traverse=True, optimize_graph=True, scheduler=None, **kwargs)
...
+# Circular reference breaker used by _normalize_seq_func.
+# This variable is recreated anew every time you call tokenize(). Note that this means
+# that you could call tokenize() from inside tokenize() and they would be fully
+# independent.
+#
+# It is a map of {id(obj): (<first seen incremental int>, obj)} which causes an object
+# to be tokenized as ("__seen", <incremental>) the second time it's encountered while
+# traversing collections. A strong reference to the object is stored in the context to
+# prevent ids from being reused by different objects.
+_seen: ContextVar[dict[int, tuple[int, object]]] = ContextVar("_seen")
```
Comment (fjetter) [inline dask/base.py:1055]: This seems overly complex. Why isn't a single `ContextVar` sufficient?

Comment (fjetter) [reply dask/base.py:1055]: I only noticed now that this was already introduced in https://github.com/dask/dask/pull/10876

File: dask/base.py [hunk 2]
```diff
@@ -1359,19 +1397,7 @@ def normalize_array(x):
...
-                except Exception:
-                    # pickling not supported, use UUID4-based fallback
-                    if not config.get("tokenize.ensure-deterministic"):
-                        data = uuid.uuid4().hex
-                    else:
-                        raise RuntimeError(
-                            f"``np.ndarray`` with object ``dtype`` {x!r} cannot "
-                            "be deterministically hashed. Please, see "
-                            "https://docs.dask.org/en/latest/custom-collections.html#implementing-deterministic-hashing "  # noqa: E501
-                            "for more information"
```
Comment (fjetter) [inline dask/base.py:1373]: This feels like an unrelated change. While the functionality is similar, the raised exception is a little different

File: dask/tests/test_tokenize.py [hunk 3]
```diff
@@ -31,54 +37,29 @@
...
+        # same amount of times).
+        args2, kwargs2 = cloudpickle.loads(cloudpickle.dumps((args, kwargs)))
+        args3, kwargs3 = cloudpickle.loads(cloudpickle.dumps((args, kwargs)))
+        args3, kwargs3 = cloudpickle.loads(cloudpickle.dumps((args3, kwargs3)))
 
-            if deterministic != "maybe":
-                assert (before == after) is deterministic
+        tok2 = tokenize(*args2, **kwargs2)
+        tok3 = tokenize(*args3, **kwargs3)
+        assert tok2 == tok3
```
Comment (fjetter) [inline dask/tests/test_tokenize.py:62]: I'm not sure if this is what you are referring to in the comment above but don't we want this to be equal as before?
```suggestion
        assert tok2 == tok3 == before
```

File: dask/tests/test_tokenize.py [hunk 4]
```diff
@@ -185,12 +161,11 @@ def test_tokenize_numpy_array_on_object_dtype():
...
     ) == check_tokenize(np.array([(1, "a"), (1, None), (1, "aaa")], dtype=object))
 
-    # Trigger non-deterministic hashing for object dtype
-    class NoPickle:
+    class NeedsCloudPickle:
         pass
 
-    x = np.array(["a", None, NoPickle], dtype=object)
-    check_tokenize(x, idempotent=False)
+    x = np.array(["a", None, NeedsCloudPickle], dtype=object)
```
Comment (fjetter) [inline dask/tests/test_tokenize.py:167]: Probably doesn't matter much but it's more likely to be an instance
```suggestion
    x = np.array(["a", None, NeedsCloudPickle()], dtype=object)
```

File: dask/tests/test_tokenize.py [hunk 5]
```diff
@@ -272,21 +247,82 @@ def test_tokenize_numpy_memmap_no_filename():
...
+    # Test env override
+    assert tokenize(inc, ensure_deterministic=False) != tokenize(
+        inc, ensure_deterministic=False
+    )
+    with pytest.raises(TokenizationError, match="Cannot tokenize"):
+        tokenize(inc, ensure_deterministic=True)
+
+    # Test ctx variable cleanup
+    with pytest.raises(LookupError):
+        _ensure_deterministic.get()
```
Comment (fjetter) [inline dask/tests/test_tokenize.py:293]: This test is about ufuncs. Asserting on this ctxvar feels odd.

File: dask/tests/test_tokenize.py [hunk 6]
```diff
@@ -310,42 +346,173 @@ def test_normalize_base():
...
 
 
-def test_tokenize_local_functions():
-    a, b, c, d, e = (
-        # Note: Same line, same code lambdas cannot be distinguished
+def _local_functions():
+    all_funcs = [
         lambda x: x,
         lambda x: x + 1,
-        lambda x: x,
```
Comment (fjetter) [inline dask/tests/test_tokenize.py:324]: why remove that?

File: dask/base.py (hunk 2 above)
Comment (crusaderky) [reply dask/base.py:1373]: Blocked by and incorporates https://github.com/dask/dask/pull/10883.
To review this PR before the above is merged, just look at the last commit.

File: dask/tests/test_tokenize.py (hunk 6 above)
Comment (crusaderky) [reply dask/tests/test_tokenize.py:324]: This PR incorporates https://github.com/dask/dask/pull/10883.
To review this PR before the above is merged, just look at the last commit.
This comment is about a change in #10883; please keep the critique there.

File: dask/tests/test_tokenize.py (hunk 3 above)
Comment (crusaderky) [reply dask/tests/test_tokenize.py:62]: This PR incorporates https://github.com/dask/dask/pull/10883.
To review this PR before the above is merged, just look at the last commit.
This comment is about a change in #10883; please keep the critique there.

File: dask/base.py (hunk 2 above)
Comment (crusaderky) [reply dask/base.py:1373]: This PR incorporates https://github.com/dask/dask/pull/10883.
To review this PR before the above is merged, just look at the last commit.
This comment is about a change in #10883; please keep the critique there.

File: dask/tests/test_tokenize.py (hunk 5 above)
Comment (crusaderky) [reply dask/tests/test_tokenize.py:293]: The test may fail if I were to tamper with the ctxvar without using the helper function `_maybe_raise_nondeterministic`

Comment (hendrikmakait) [reply dask/tests/test_tokenize.py:293]: This feels like it would be useful as a comment.

File: dask/tests/test_tokenize.py [hunk 7]
```diff
@@ -280,11 +301,28 @@ def __reduce__(self):
...
-        with pytest.raises(RuntimeError, match="cannot be deterministically hashed"):
+        with pytest.raises(
+            TokenizationError, match="cannot be deterministically hashed"
+        ):
             tokenize(c)
 
+    # Test env override
+    assert tokenize(c, ensure_deterministic=False) != tokenize(
+        c, ensure_deterministic=False
+    )
```
Comment (hendrikmakait) [inline dask/tests/test_tokenize.py:318]: Should this be nested within the block above?

File: dask/tests/test_tokenize.py [hunk 8]
```diff
@@ -264,13 +270,28 @@ def test_normalize_numpy_ufunc_unserializable():
...
         with pytest.raises(
-            RuntimeError, match=r"Cannot tokenize.*dask\.array\.ufunc.*instead"
+            TokenizationError, match=r"Cannot tokenize.*dask\.array\.ufunc.*instead"
         ):
             tokenize(inc)
 
+    # Test env override
+    assert tokenize(inc, ensure_deterministic=False) != tokenize(
+        inc, ensure_deterministic=False
+    )
```
Comment (hendrikmakait) [inline dask/tests/test_tokenize.py:287]: Should this be nested within the block above?

File: dask/tests/test_tokenize.py [hunk 9]
```diff
@@ -308,18 +346,58 @@ def test_normalize_base():
...
+            normalize_token(o)
 
-    # Two objects don't tokenize to the same token even if their pickle output is
-    # identical. Stress id reuse by creating and dereferencing many objects in quick
-    # succession.
-    assert len({tokenize(object()) for _ in range(100)}) == 100
+    # Test env override
+    assert tokenize(o, ensure_deterministic=False) == tokenize(
+        o, ensure_deterministic=False
+    )
```
Comment (hendrikmakait) [inline dask/tests/test_tokenize.py:374]: Should this be nested within the block above?

File: dask/tests/test_tokenize.py (hunk 5 above)
Comment (crusaderky) [reply dask/tests/test_tokenize.py]: I reworked it.

File: dask/tests/test_tokenize.py (hunk 8 above)
Comment (crusaderky) [reply dask/tests/test_tokenize.py:299]: No. These 3 tests use the same pattern:

```python
with dask.config.set({"tokenize.ensure-deterministic": False}):
    tokenize(...)
with dask.config.set({"tokenize.ensure-deterministic": True}):
    tokenize(...)
tokenize(..., ensure_determinsitic=False)
tokenize(..., ensure_deterministic=True)
```
This makes them impervious to whatever the config is set to globally.

File: dask/tests/test_tokenize.py (hunk 7 above)
Comment (crusaderky) [reply dask/tests/test_tokenize.py:326]: same answer as above

File: dask/tests/test_tokenize.py (hunk 9 above)
Comment (crusaderky) [reply dask/tests/test_tokenize.py:378]: same answer as above

**Comment by crusaderky** [comment]
@hendrikmakait @fjetter all comments have been addressed

File: dask/tests/test_tokenize.py (hunk 8 above)
Comment (hendrikmakait) [reply dask/tests/test_tokenize.py:299]: My bad, I attributed the inline comment only to this test case, not the one below.

Comment (hendrikmakait) [reply dask/tests/test_tokenize.py:299]: I also wasn't sure if this should test overriding "whatever" the config is set to globally or overriding _any possible_ config value.

**Review by hendrikmakait — APPROVED** [review]
Thanks, @crusaderky! LGTM assuming CI is green

**Pull request merged**